flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveTableCompactSinkITCase.java

```diff
@@ -0,0 +1,250 @@
+                        + " 'compaction.small-files.avg-size' = '9b', "
+                        + " 'sink.parallelism' = '4'" // set sink parallelism = 8
+                        + ")");
+
+        tableEnv.executeSql(
+                        "insert into src values ('k1', 'v1', 1), ('k2', 'v2', 1),"
+                                + "('k3', 'v3', 2), ('k4', 'v4', 2), ('k5', 'v5', 1)")
+                .await();
+
+        // one row is 6 bytes, so the partition "p=2" will contain two files, one of which only
+        // contain one row, the average size is 6 bytes. so the files should be merged to one single
+        // file.
+        List<Path> files = listDataFiles(Paths.get(warehouse, "src/p=2"));
+        assertThat(files).hasSize(1);
+
+        // the partition "p=1" will contain two files, one contain two rows, and the other contain
+        // one row. the average size is 9 bytes, so the files shouldn't be merged
+        files = listDataFiles(Paths.get(warehouse, "src/p=1"));
+        assertThat(files).hasSize(2);
+
+        List<String> result = toSortedResult(tableEnv.executeSql("select * from src"));
+        assertThat(result.toString())
+                .isEqualTo(
+                        "[+I[k1, v1, 1], +I[k2, v2, 1], +I[k3, v3, 2], +I[k4, v4, 2], +I[k5, v5, 1]]");
+    }
+
+    @Test
+    public void testCompactInNoBlockingMode() throws Exception {
+        // create a environment with non-blocking mode
+        tableEnv = createNoBlockingModeTableEnv();
+        tableEnv.executeSql(
+                "CREATE TABLE src ("
+                        + " key string,"
+                        + " value string"
+                        + ") partitioned by (p int) TBLPROPERTIES ("
+                        + " 'auto-compaction' = 'true', "
+                        + " 'compaction.small-files.avg-size' = '9b', "
+                        + " 'sink.parallelism' = '4'" // set sink parallelism = 8
+                        + ")");
+        assertThatThrownBy(
+                        () ->
+                                tableEnv.executeSql(
+                                        "insert into src values ('k1', 'v1', 1), ('k2', 'v2', 1),"
+                                                + "('k3', 'v3', 2), ('k4', 'v4', 2), ('k5', 'v5', 1)"))
+                .isInstanceOf(UnsupportedOperationException.class)
+                .hasMessage(
+                        String.format(
+                                "Auto compaction for Hive sink in batch mode is not supported when the %s is not %s.",
+                                ExecutionOptions.BATCH_SHUFFLE_MODE.key(),
+                                BatchShuffleMode.ALL_EXCHANGES_BLOCKING));
+    }
+
+    private TableEnvironment createNoBlockingModeTableEnv() {
+        EnvironmentSettings settings = EnvironmentSettings.inBatchMode();
+        settings.getConfiguration()
+                .set(ExecutionOptions.BATCH_SHUFFLE_MODE, BatchShuffleMode.ALL_EXCHANGES_PIPELINED);
```

Comment:
I think this option is job level, so we don't need to create a new TableEnvironment.
```
        tableEnv.getConfig().set(ExecutionOptions.BATCH_SHUFFLE_MODE, BatchShuffleMode.ALL_EXCHANGES_PIPELINED);
```